Look for additional opportunities for stack allocation

hzongaro opened the pull request:
Look for additional opportunities for stack allocation
    
Check for cases in which locally allocated objects are only copied to other local variables, without values ever escaping.
    
These escape analysis enhancements were prototyped by Vijay Sundarsen @vijaysun-omr and Mark Stoodley @mstoodle .
    
Fixes:  #2072
    
Signed-off-by:  Henry Zongaro <zongaro@ca.ibm.com>

charliegracie commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -155,7 +155,8 @@ bool TR_EscapeAnalysis::isImmutableObject(TR::Node *node)
       return false;
 
-   if (!strncmp("java/lang/Integer", getClassName(node->getFirstChild()), 17))
+   if (!strncmp("java/lang/Integer", getClassName(node->getFirstChild()), 17) ||
```
Also what about Float, Long, etc.... Seems like future proofing now to handle all of the primitive wrapper classes might be a good idea.

fjeremic commented on runtime/compiler/optimizer/EscapeAnalysis.hpp:
```diff
@@ -568,6 +576,8 @@ class TR_EscapeAnalysis : public TR::Optimization
    TR_BitVector              *_blocksWithFlushOnEntry;
    TR_BitVector              *_visitedNodes;
+   TR_BitVector              *_aliasesOfAllocNode;
+   TR_BitVector              *_aliasesOfAllocNode2;
```
This does not seem like a good field name. Could we name it better? Can we add documentation on what this field does exactly? Similarly for others, i.e. `candidate2`

hzongaro replied:
@fjeremic Filip, sorry for the long delay in responding.  How about instead of `candidate2` and `_aliasesOfAllocNode2`, we used names like `otherAllocNode` and `_aliasesOfOtherAllocNode`, along with a repetition of Vijay's inline comment describing `_aliasesofAllocNode`?  Something like this:

```
/** 
 * Contains sym refs that are just aliases for a fresh allocation
 * i.e., it is used to track allocations in cases such as
 * ...
 * a = new A() 
 * ...
 * b = a 
 * ...
 * c = b 
 *    
 * In this case a, b and c will all be considered aliases of an alloc node and
 * so a load of any of those sym refs will be treated akin to how the fresh
 * allocation would be treated
 */ 
   TR_BitVector              *_aliasesOfAllocNode;
/**
 * Contains sym refs that are just aliases for a second fresh allocation
 * that is under consideration, as with @ref _aliasesOfAllocNode
 */
   TR_BitVector              *_aliasesOfOtherAllocNode;
```

I know the name _aliasesOfOtherAllocNode doesn't really convey much more about its purpose than _aliasesOfAllocNode2 did, but they're really two vectors tracking aliases of particular candidate allocations that are under consideration at any one point in time.  Do you have any other suggestions?

Perhaps more importantly, we could add a comment over the loop that iterates over `cursor2` - something like this:

```
      // Loop over definitions for this use and over all the candidate allocations
      // If the definition comes directly from a candidate for stack allocation, it's harmless;
      // if it's copied from a variable that's aliased with a candidate for stack allocation that
      // was allocated in the same block, it's harmless
      for (cursor2.SetToFirstOne(); cursor2.Valid(); cursor2.SetToNextOne())
               {
               int32_t defIndex2 = cursor2;
               if (defIndex2 == 0)
                  {
                  allnewsonrhs = false;
                  break;
                  }

               TR::Node *defNode2 = _useDefInfo->getNode(defIndex2);
               TR::Node *firstChild = defNode2->getFirstChild();
               bool rhsIsHarmless = false;
               for (Candidate *candidate2 = _candidates.getFirst(); candidate2; candidate2 = candidate2->getNext())
                  {
                  if (candidate2->_node == firstChild)
                     {
                     rhsIsHarmless = true;
                     break;
```

Does that help?

@vijaysun-omr Vijay, does that capture things accurately?  Any other thoughts?

hzongaro replied:
Via private communication, Vijay agreed with those suggested adjustments to variable names and comments.

0xdaryl commented:
This PR will ultimately have a more descriptive commit message, won't it?

hzongaro commented:
> This PR will ultimately have a more descriptive commit message, won't it?

Vijay @vijaysun-omr , Daryl @0xdaryl, I've finally come back to this PR, squashed the various commits and tried to compose a better description of the changes.  Does that describe things adequately?  Once you're OK with the commit messages, I'll remove the _WIP_ tag.

fjeremic commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -155,7 +155,17 @@ bool TR_EscapeAnalysis::isImmutableObject(TR::Node *node)
+   if (!strncmp("java/lang/", className, 10) &&
+          (!strcmp("Integer", &className[10]) ||
+           !strcmp("Long", &className[10]) ||
+           !strcmp("Short", &className[10]) ||
```
Is there really not a better way? Not saying we should fix this here but this looks like a really bad way to looking for particular classes. I know for recognized methods we have things like `TR::java_util_HashMap_all` to capture all HashMap methods. Wondering if something similar exists for classes.

hzongaro replied:
@fjeremic Filip, I'm trying to figure out whether there's a cleaner way of doing this check.

hzongaro replied:
@fjeremic Filip, I don't see any existing mechanism for doing what you're suggesting.  Would you like me to open an issue to track that as an improvement?

fjeremic replied:
Yes, let's open up an issue. I see similar types of `strcmp` code in other parts of the code base. I do think we need some notion on "recognized classes" to avoid this type of cruft. Thanks!

hzongaro replied:
Opened issue #3804

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -2332,6 +2432,32 @@ bool TR_EscapeAnalysis::checkOverlappingLoopAllocation(TR::Node *node, TR::Node
 
 
+
+void TR_EscapeAnalysis::collectAliasesOfAllocations(TR::Node *node, TR::Node *allocNode, bool collectAliases)
```
I am not a fan of this bool parameter that makes this method just do the visit - basically putting all of the logic in an if... I would rather we had a method to just visit and a method to collect aliases and dropped this boo parameter which is going to make usage harder to understand/follow.

hzongaro replied:
Andrew @andrewcraik, Just to be sure I'm clear on what you're suggesting, do you want to see a `visitTree` method that just visits nodes in a tree and a `collectAliasesOfAllocations` that both visits and collects aliases unconditionally?  Or do you want to see a `visitTree` method that visits nodes and also takes a function pointer (or an object) that it calls at each node, where that function would either be no-op or collect aliases?

andrewcraik replied:
for this I'd suggest the former - a `visitTree` that unconditionally visits the tree and marks it as visited and a `collectAliasesofAllocations` that does what it currently does when the Boolean is true. Function pointers can make debugging the compiler harder so I try to avoid them.

hzongaro replied:
Split `collectAliasesOfAllocations` into two methods in commit 665010adf18731078ee8181b05cf4ae0480c1a69, as Andrew suggested

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -2332,6 +2432,32 @@ bool TR_EscapeAnalysis::checkOverlappingLoopAllocation(TR::Node *node, TR::Node
+
+   if (collectAliases && node->getOpCode().isStore() && node->getSymbol()->isAutoOrParm())
+      {
+      if (node->getFirstChild() == allocNode)
```
code style - this if ladder should have {}s IMO just due to complexity. Given teh complexity of the conditions here can we put them one per line with the && at the star to the line followed by teh condition? This will reduce the line width and make this much more clear and easier to reason about. This is a style used elsewhere in the compiler when dealing with such complex conditions and has tended to work well.

hzongaro replied:
Cleaned this up in commit 414b74b10957a8ae1b8a4426f0c29dec2a905275

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -2314,7 +2400,21 @@ bool TR_EscapeAnalysis::checkOverlappingLoopAllocation(TR::Node *node, TR::Node
+         _aliasesOfAllocNode->reset(node->getSymbolReference()->getReferenceNumber());
+      }
+
+   if ((_valueNumberInfo->getValueNumber(node) == _valueNumberInfo->getValueNumber(useNode)) &&
```
for efficiency perhaps do the node != allocNode first since it is cheaper?

hzongaro replied:
Cleaned this up in commit 414b74b10957a8ae1b8a4426f0c29dec2a905275

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -2202,17 +2252,49 @@ bool TR_EscapeAnalysis::checkOtherDefsOfLoopAllocation(TR::Node *useNode, Candid
+                           trackAliases = true;
+                       }
+
+                    if (/* (_valueNumberInfo->getValueNumber(defNode2) == _valueNumberInfo->getValueNumber(otherAllocNode->_node)) && */
```
please remove dead code rather than commenting it out

hzongaro replied:
Removed dead code in 414b74b10957a8ae1b8a4426f0c29dec2a905275

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -2178,18 +2205,41 @@ bool TR_EscapeAnalysis::checkOtherDefsOfLoopAllocation(TR::Node *useNode, Candid
+
       bool allnewsonrhs = false;
-      if ((_valueNumberInfo->getValueNumber(defNode) == _valueNumberInfo->getValueNumber(candidate->_node)))
+      //if ((_valueNumberInfo->getValueNumber(defNode) == _valueNumberInfo->getValueNumber(candidate->_node)))
```
again remove dead code. Once we agree on the change a change to remove the {}s and reduce the indentation is probably also needed.

hzongaro replied:
Removed dead code and braces, and adjusted indentation in commit 414b74b10957a8ae1b8a4426f0c29dec2a905275

andrewcraik commented:
Overall, I think this set of changes needs much better trace. There is nothing printed to the trace logs about what nodes are considered as equivalent in the new scheme or what the analysis is doing. When trace is enabled we need to be able to follow the analysis from the trace for debug purposes since we may only ever get the log as the means of finding and fixing a problem.

Additionally, is there any reasonable way to have an env var or other mechanism to disable this new opt to aid debugging? If we find failures after this is delivered it would be nice to have an easy way to rule this change in or out.

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -155,7 +155,17 @@ bool TR_EscapeAnalysis::isImmutableObject(TR::Node *node)
       return false;
 
-   if (!strncmp("java/lang/Integer", getClassName(node->getFirstChild()), 17))
+   char *className = getClassName(node->getFirstChild());
```
getClassName may not be new, but I think there needs to be a safety check that the node is one which should have a class as its symbol and that we can get the name. We should handle this returning NULL in cases where it can't be gotten.

hzongaro replied:
Thanks, Andrew @andrewcraik.  I've been pawing around trying to figure out the right way to perform the check you're recommending.  Would this be the right way?

```
if (classNode->getOpCode().hasSymbolReference())
   {
   TR::SymbolReference *symRef = classNode->getSymbolReference();
   TR::Symbol *sym = symRef->getSymbol();
   if (sym->isClassObject() && !sym->addressIsCPIndexOfStatic())
      ...
   }
```

andrewcraik replied:
So the child node will need to be a loadaddr otherwise the symref won't be describing the thing we want (currently news have a loadaddr as a child, but this will be changing). So the first part of the check needs to be that we have a loadaddr.

Usage elsewhere in the compiler suggests sym->isClassObject() is the check you need. I don't know that you need the addressIsCPIndexOfStatic.

hzongaro replied:
Added checking to `getClassName` and its callers in commit f12434d607be9a31b4674ea1c8bcedceb82558da

hzongaro commented:
> Overall, I think this set of changes needs much better trace. There is nothing printed to the trace logs about what nodes are considered as equivalent in the new scheme or what the analysis is doing. When trace is enabled we need to be able to follow the analysis from the trace for debug purposes since we may only ever get the log as the means of finding and fixing a problem.

Andrew @andrewcraik, I have a commit pending that adds additional tracing, though I'll need someone to review for use of terminology, etc.

> Additionally, is there any reasonable way to have an env var or other mechanism to disable this new opt to aid debugging? If we find failures after this is delivered it would be nice to have an easy way to rule this change in or out.

I'm working on using a `TR_disableEALoopAliasAllocationChecking` environment variable for disabling the additional alias checking.  I'll run complete testing with it enabled and disabled.

I was a bit torn on whether to weave checking of the associated bool, `_doLoopAliasAllocationChecking`, throughout the `if` that calculates `allnewsonrhs` with the old and new logic intermingled, but ended up deciding to simply pull an old version of that logic and a new version into new methods for easier maintenance.  Commit pending.

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -1489,7 +1527,10 @@ Candidate *TR_EscapeAnalysis::createCandidateIfValid(TR::Node *node, TR_OpaqueCl
                {
-               printf("secs Class %s implements Runnable in %s\n", getClassName(classNode), comp()->signature());
+               const char *className = getClassName(classNode->getSecondChild());
+               printf("secs Class %s implements Runnable in %s\n",
```
please convert to guarded trace

hzongaro replied:
@andrewcraik Andrew, are you asking me to replace the `printf` with `traceMsg`?  If so, there are a few other places with calls to `printf` that I'll take care of as well.

andrewcraik replied:
yes please - printf is not a nice way to debug anything in the JIT so we should only commit code that logs using the traceMsg mechanism.

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -1382,13 +1416,17 @@ void TR_EscapeAnalysis::findCandidates()
             {
-            traceMsg(comp(), "Found [%p] new %s\n", node, getClassName(node->getFirstChild()));
+            const char *className = getClassName(node->getFirstChild());
+            traceMsg(comp(), "Found [%p] new %s\n", node,
```
probably deserves an if(trace())

hzongaro replied:
That one is actually in an `if (trace())`:
```
      if (trace())
         {
         if (node->getOpCodeValue() == TR::New)
            {
            const char *className = getClassName(node->getFirstChild());
            traceMsg(comp(), "Found [%p] new %s\n", node,
                     className ? className : "<Missing class name>");
            }
         else if (node->getOpCodeValue() == TR::newarray)
```

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -1382,13 +1416,17 @@ void TR_EscapeAnalysis::findCandidates()
             {
-            traceMsg(comp(), "Found [%p] anewarray %s\n", node, getClassName(node->getSecondChild()));
+            const char *className = getClassName(node->getSecondChild());
+            traceMsg(comp(), "Found [%p] anewarray %s\n", node,
```
similar trace guard

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -2178,69 +2231,134 @@ bool TR_EscapeAnalysis::checkOtherDefsOfLoopAllocation(TR::Node *useNode, Candid
+
+            TR::Node *defNode2 = _useDefInfo->getNode(defIndex2);
+            TR::Node *firstChild = defNode2->getFirstChild();
+            bool rhsIsHarmless = false;
```
could we have a comment on what harmless means in this context to provide the intuition for all the checks that follow?

hzongaro replied:
Addressed this in commit aa48bf3fd7a279f3feb25aba345b7ff6850f3c8f

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -2178,69 +2231,134 @@ bool TR_EscapeAnalysis::checkOtherDefsOfLoopAllocation(TR::Node *useNode, Candid
                         }
+
+                     if (fieldName && (fieldNameLen > 0) &&
+                         !strncmp(fieldName, "java/lang/Integer$IntegerCache.cache", 36))
```
does long have an equivalent cache that should be treated the same way for symmetry?

hzongaro replied:
It seems it does, as do `java.lang.Byte` and `java.lang.Character`

hzongaro replied:
Expanded the set of caches checked for in commit 92b0cb068aa2ec511d8df1d4cc8a071eac10ca4d

hzongaro commented:
Andrew @andrewcraik, I think I've addressed all your comments.  I would appreciate it if you could have a look - particularly at those that address your [review comment on `rhsIsHarmless`](https://github.com/eclipse/openj9/pull/2752/#pullrequestreview-182719656) and your [request for additional tracing and a way of disabling the new code](https://github.com/eclipse/openj9/pull/2752/#issuecomment-444921956).

Once you're OK with these changes and any follow on changes, I can collapse these all down to one (or two) commits.

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -2432,8 +2432,14 @@ bool TR_EscapeAnalysis::checkAllNewsOnRHSInLoopWithAliasing(int32_t defIndex, TR
-                      !strncmp(fieldName, "java/lang/Integer$IntegerCache.cache", 36))
+                  if (fieldName && (fieldNameLen > 10) &&
+                        !strncmp("java/lang/", fieldName, 10) &&
+                          (!strcmp("Integer$IntegerCache.cache", &fieldName[10]) ||
```
I think we need to use `strncmp` here because names are not guaranteed to be null terminated.

andrewcraik replied:
Otherwise - I think this looks good.

hzongaro replied:
OK.  Now I see `staticName` yields the name and type signature, separated by a space.  I'll fix that.

hzongaro replied:
Corrected that blunder in commit 4bf7ad3bb06a85e54e7d8562d875ffc474c2f8fd

hzongaro commented:
> I think this might be more clear as disableLoopAliasAllocationChecking == NULL

Fixed in commit 4bf7ad3bb06a85e54e7d8562d875ffc474c2f8fd
@andrewcraik

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -521,6 +556,12 @@ int32_t TR_EscapeAnalysis::performAnalysisOnce()
       _visitedNodes = new (trStackMemory()) TR_BitVector(comp()->getNodeCount(), trMemory(), stackAlloc, growable);
+      _aliasesOfAllocNode =
+          _doLoopAllocationAliasChecking
+                 ? new (trStackMemory()) TR_BitVector(comp()->getSymRefCount(), trMemory(), stackAlloc, growable) : NULL;
```
Adjusted in commit 4bf7ad3bb06a85e54e7d8562d875ffc474c2f8fd

andrewcraik commented:
Alrighty I'm good with this. @hzongaro can you squash to the final form of commits you expect and @vijaysun-omr to see if he has any concerns? we can then sanity ready for merge.

hzongaro commented:
Squashed changes down to three commits:

- 4ee42c01289eca51556948f197b4985d5145b596
- 7d903f9b4c8c070809f0764f3b276876d63904a4
- 7d903f9b4c8c070809f0764f3b276876d63904a4

charliegracie commented:
@andrewcraik and @vijaysun-omr is this ready to go now?

vijaysun-omr commented:
I'm fine with this now. I believe @hzongaro was planning to check with Andrew before this goes in.

andrewcraik commented:
Jenkins test sanity xlinux,win.plinux jdk8,jdk11

andrewcraik commented:
Jenkins test sanity xlinux,win,plinux jdk8,jdk11
